### PR TITLE
Major overhaul with completely different architecture

### PR DESCRIPTION
**Pull request by Toflar**
This PR contains a major overhaul of the whole library.
I've focussed on making Escargot a general library that can be used to crawl whatever you want to crawl based on HTTP, not just websites. The core `Escargot` instance is now pretty small providing four basic events that allow to build a custom crawler.

The website crawling part is now extracted into own subscribers. I've also introduced PSR-3 logging instead of additional events one can use to log.

For more details, read the updated `README`.

TODOs:

- [x] Unit tests for the `MaxDepthSubscriber`
- [x] Unit tests for the `RobotsSubscriber`

### CODE REVIEW

File: src/Escargot.php [hunk 1]
```diff
@@ -306,37 +285,51 @@ public function crawl(): void
...
-                        // TODO: another event?
-                        --$this->runningRequests;
-                        $response->cancel();
-                        continue;
-                    }
+                // Dispatch event
+                $this->getEventDispatcher()->dispatch(new ResponseEvent($this, $response, $chunk));
+
+                // Response was canceled by listener
+                if ('Response has been canceled.' === $response->getInfo('error')) {
```
Comment (Toflar) [inline src/Escargot.php]: This here @nicolas-grekas :)

Comment (nicolas-grekas) [reply src/Escargot.php]: /cc @seldaek we talked about that recently.
@Toflar I'd accept a PR to add a "canceled" info in the contract and in the component if you're up to.

Comment (Toflar) [reply src/Escargot.php]: Sure! Will submit next week. master or 4.3?

**Comment by nicolas-grekas** [comment]
4.4

**Comment by Toflar** [comment]
Ah, I keep forgetting that `master` is 5.0 already 😄 Here we go: https://github.com/symfony/symfony/pull/34044
Further discussions over at the Symfony repository. Thanks for taking the time, once again Nicolas!

**Pull request merged**